torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1075,14 +1025,58 @@ void TensorPipeAgent::markFutureWithError(
                      futureMessage{std::move(futureMessage)},
                      errorMsg{std::move(errorMsg)}]() mutable {
       futureMessage->futMsg.setError(std::move(errorMsg));
-      // The future's callbacks may schedule further RPCs, increasing the count.
-      // Thus we must decrease it after completing the future, otherwise it may
-      // briefly dip to zero and trick join into thinking all work is done.
-      decreaseCallCount(clientActiveCalls_);
     });
   }
 }
 
+std::vector<c10::DeviceIndex> TensorPipeAgent::getDevicesForTensors(
+    const std::string& remoteName,
+    const Message& message) const {
+  const auto& deviceMaps =
+      message.isRequest() ? opts_.deviceMaps : reverseDeviceMaps_;
+  const auto messageType = message.isRequest() ? "Request" : "Response";
+
+  const auto& iter = deviceMaps.find(remoteName);
+  if (iter == deviceMaps.end()) {
+    for (const auto& t : message.tensors()) {
+      TORCH_CHECK(
+        t.device().is_cpu(),
+        "TensorPipe RPC backend only supports CPU tensors by default, please "
+        "move your tensors to CPU before sending them over RPC, or call "
+        "`set_device_map` on `TensorPipeRpcBackendOptions` to explicitly "
+        "configure device mapping. ",
+        messageType,
+        " device mapping is not available for destination ",
+        remoteName,
+        ", but found tensor on device: ",
+        t.device());
+    }
+    return {};
+  } else {
+    std::vector<c10::DeviceIndex> deviceIndices;
+    deviceIndices.reserve(message.tensors().size());
+    const auto& deviceMap = iter->second;
+    for (const auto& t : message.tensors()) {
+      const auto deviceIter = deviceMap.find(t.device().index());
+      if (deviceIter == deviceMap.end()) {
+        TORCH_CHECK(
+          t.device().is_cpu(),
+          "TensorPipe RPC backend only supports CPU tensors by default. ",
+          messageType,
+          " device mapping is not available for destination ",
+          remoteName,
+          " for device ",
+          t.device(),
+          " but received a tensor on that device.");
```

Comment:
There are two nearly-identical instances of `TORCH_CHECK(t.device().is_cpu(), ...)` and there is also the `checkCPUTensor` function that does the same thing. Is this last function even used anymore? Could we consolidate this?

Reply:
yes, should have removed `checkCPUTensor`.